charms/reactive/endpoints.py

```diff
@@ -377,8 +377,12 @@ def application_name(self):
 
             relation.units[0].unit_name.split('/')[0]
         """
-        if self._application_name is None and self.units:
-            self._application_name = self.units[0].unit_name.split('/')[0]
+        if self._application_name is None:
+            if self.units:
+                self._application_name = self.units[0].unit_name.split('/')[0]
+            else:
+                # NOTE(ganso): This is a single unit peer relation corner case
+                return hookenv.application_name()
```

Comment:
is there any harm in assigning this to self._application_name to avoid repeated lookups during a single hook execution?

Reply:
hey James. I had left it as it is because right now it is addressing a problem that occurs only when there is a single cluster unit, which is rarely seen on production deployments, so the performance impact is not that significant. My concern is that we are introducing a change in the library that is used by every charm, but right now we are testing just a single charm with this updated code, we don't fully know the impact of the change on other charms or other corner cases we don't know about. What I suspect could be a negative impact of assigning the value to the attribute is that if there is a window of opportunity that happens in other charms or other corner cases we don't know about (such as maybe a relation-joined hook running concurrently with something else), with the original code and proposed fix, that problem could possibly correct itself as part of the hook execution, as the attribute would continue to be None and re-evaluate next time the code is reached (which it does on almost every reactive handle). If we assign the value, then it will not be re-evaluated anymore. On the other hand, if it causes a problem it will allow us to spot and fix them (hopefully not after it is running in production). Overall, IMO, the risk/benefit isn't worth it, but I'm ok if you prefer to have it. I have made the change and tested in my lab and it works fine, but only tested this specific scenario.